pkg/virt-api/rest/streamer_test.go

```diff
@@ -0,0 +1,407 @@
+				fetchVMICalled = true
+				return testVMI, nil
+			},
+			validateVMI: func(vmi *v1.VirtualMachineInstance) *errors.StatusError {
+				validateVMICalled = true
+				return nil
+			},
+			dial: func(vmi *v1.VirtualMachineInstance) (net.Conn, *errors.StatusError) {
+				dialCalled = true
+				return serverConn, nil
+			},
+			streamToClient: func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+				result <- nil
+				streamToClientCalled <- struct{}{}
+			},
+			streamToServer: func(clientSocket *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult) {
+				result <- nil
+				streamToServerCalled <- struct{}{}
+			},
+		}
+
+		httpReq = httptest.NewRequest(http.MethodGet, fmt.Sprintf("/apis/subresources.kubevirt.io/v1alpha3/namespaces/%s/virtualmachineinstances/%s/ssh/22", testNamespace, testName), nil)
+		req = restful.NewRequest(httpReq)
+		req.Request.URL.Scheme = "wss"
+		respRecorder = httptest.NewRecorder()
+		resp = restful.NewResponse(respRecorder)
+
+		fetchVMICalled = false
+		validateVMICalled = false
+		dialCalled = false
+	})
+	It("fetches a VirtualMachineInstance", func() {
+		streamer.Handle(req, resp)
+		Expect(fetchVMICalled).To(BeTrue())
+	})
+	It("fetches the VMI specified in the request params", func() {
+		params := req.PathParameters()
+		params[NamespaceParamName] = testNamespace
+		params[NameParamName] = testName
+
+		streamer.Handle(req, resp)
+		Expect(fetchVMICalled).To(BeTrue())
+	})
+	It("validates the VMI", func() {
+		streamer.Handle(req, resp)
+		Expect(validateVMICalled).To(BeTrue())
+	})
+	It("validates the fetched VMI", func() {
+		streamer.validateVMI = func(vmi *v1.VirtualMachineInstance) *errors.StatusError {
+			Expect(vmi).To(Equal(testVMI))
+			return nil
+		}
+		streamer.Handle(req, resp)
+	})
+	It("does not validate the VMI if it can't be fetched", func() {
+		streamer.fetchVMI = func(_, _ string) (*v1.VirtualMachineInstance, *errors.StatusError) {
+			return nil, errors.NewInternalError(goerrors.New("test error"))
+		}
+
+		Expect(streamer.Handle(req, resp)).To(HaveOccurred())
+		Expect(validateVMICalled).To(BeFalse())
+	})
+	It("dials the VMI", func() {
+		streamer.Handle(req, resp)
+		Expect(dialCalled).To(BeTrue())
+	})
+	It("dials the fetched VMI", func() {
+		streamer.dial = func(vmi *v1.VirtualMachineInstance) (net.Conn, *errors.StatusError) {
+			Expect(vmi).To(Equal(testVMI))
+			return nil, nil
+		}
+		streamer.Handle(req, resp)
+	})
+	It("does not dial when VMI is invalid", func() {
+		streamer.validateVMI = func(_ *v1.VirtualMachineInstance) *errors.StatusError {
+			return errors.NewInternalError(goerrors.New("test error"))
+		}
+
+		Expect(streamer.Handle(req, resp)).To(HaveOccurred())
+		Expect(dialCalled).To(BeFalse())
+	})
+	It("upgrades the client connection", func() {
+		srv, ws, wsResp, err := testWebsocketDial(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
```

Comment:
This is used a lot in the following tests and could be moved around to deduplicate a little.